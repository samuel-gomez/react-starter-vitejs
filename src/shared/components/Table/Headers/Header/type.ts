import { ReactNode } from 'react';

export type THeader = {
  label: string;
  key: string;
  field?: string;
  infobulle?: ReactNode;
};
