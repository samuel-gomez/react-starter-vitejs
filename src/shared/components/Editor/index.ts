export { withEditor, useEditable, EditorHeader } from './Editor';
export type { Tknobs, TReturnUseToggleEditor, TonChange } from './Editor';
