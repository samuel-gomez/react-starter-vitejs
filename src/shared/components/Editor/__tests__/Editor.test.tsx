import { Text } from '@axa-fr/react-toolkit-all';
import { act, render, renderHook } from '@testing-library/react';
import { type ComponentPropsWithoutRef, type FocusEvent } from 'react';
import type { TEvent } from 'shared/types.d';
import { LabelEditor, mergePropsAndKnobs, setValue, useEditable, useToggleEditor, withEditor } from '../Editor';

describe('setValue', () => {
  it.each`
    value      | expected
    ${'true'}  | ${false}
    ${'false'} | ${true}
    ${'value'} | ${'value'}
  `('Should return expected: $expected when value: $value', ({ value, expected }) => {
    const result = setValue(value);
    expect(result).toEqual(expected);
  });
});

describe('mergePropsAndKnobs', () => {
  const otherProps = {
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
  };
  const propsMock = {
    ...otherProps,
    classModifier: 'error',
  };
  const knobsMock = {
    classModifier: {
      value: '',
      options: [
        {
          label: 'Success',
          value: 'success',
        },
      ],
    },
  };
  const expectedMerge = { ...otherProps, classModifier: { ...knobsMock.classModifier, value: 'error' } };

  it.each`
    props        | knobs        | expected
    ${{}}        | ${{}}        | ${{}}
    ${propsMock} | ${knobsMock} | ${expectedMerge}
  `('Should return expected: $expected when props: $props, knobs: $knobs', ({ props, knobs, expected }) => {
    const result = mergePropsAndKnobs({ props, knobs });
    expect(result).toEqual(expected);
  });
});

describe('useEditable', () => {
  const initialState = {
    classModifier: '',
    className: 'af-btn',
    label: 'valider',
    disabled: false,
    icon: '',
    autoFocus: false,
    values: [],
  };

  it('Should logEventFn have been called when onClick have been called', () => {
    const logEventFn = vi.fn();
    const { result } = renderHook(() => useEditable({ initialState, logEventFn }));
    const res = result.current.onClick('name')({ id: 'id' });
    expect(res).toEqual({ name: 'name', e: { id: 'id' } });
  });

  it('Should setValueFn have been called and return updated state value when onChange have been called with event with value', () => {
    const setValueFn = vi.fn().mockReturnValue('setted value');
    const { result } = renderHook(() => useEditable({ initialState, setValueFn }));
    act(() => result.current.onChange('classModifier')({ value: 'newvalue' }));
    expect(setValueFn).toHaveBeenCalledWith('newvalue');
    expect(result.current.state.classModifier).toEqual('setted value');
  });

  it('Should setValueFn have been called with empty string and return updated state value when onChange have been called with event with undefined value', () => {
    const setValueFn = vi.fn().mockReturnValue('setted value');
    const { result } = renderHook(() => useEditable({ initialState, setValueFn }));
    act(() => result.current.onChange('classModifier')({ value: undefined }));
    expect(setValueFn).toHaveBeenCalledWith('');
    expect(result.current.state.classModifier).toEqual('setted value');
  });

  it('Should return updated state values when onChange have been called with event with values', () => {
    const { result } = renderHook(() => useEditable({ initialState }));
    act(() => result.current.onChange('values')({ values: ['newvalue'] }));

    expect(result.current.state.values).toEqual(['newvalue']);
  });

  it('Should set autofocus state to false when onBlur have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: true } }));
    act(() => result.current.onBlur());
    expect(result.current.state.autoFocus).toBeFalsy();
  });

  it('Should set autofocus state to true when onFocus have been called', () => {
    const { result } = renderHook(() => useEditable({ initialState: { autoFocus: false } }));
    const setSelectionRange = vi.fn();
    const event = {
      preventDefault: vi.fn(),
      target: { value: '<p>the-new-value</p>', type: 'textarea', setSelectionRange },
    } as unknown as FocusEvent<HTMLInputElement>;

    act(() => result.current.onFocus(event));
    expect(result.current.state.autoFocus).toBeTruthy();
    expect(setSelectionRange).toHaveBeenCalledWith(20, 20);
  });
});

describe('withEditor', () => {
  type Props = { className?: string; onChange: ReturnType<typeof useEditable>['onChange'] };
  const onChange = vi.fn().mockImplementation(name => (e: TEvent) => {
    return { onChange: { e, name } };
  });
  const Component = ({ className, onChange: onChangeComponent }: Props) => (
    <Text
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
      onChange={onChangeComponent as unknown as ComponentPropsWithoutRef<typeof Text>['onChange']}
      className={className}
    />
  );

  it('Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true', () => {
    const useToggleEditorFn = vi.fn().mockReturnValue({
      closeEditor: vi.fn(),
      openEditor: vi.fn(),
      isOpenEditor: true,
    });
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFn);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render Component with FormEditor when apply withEditor HOC', () => {
    const ComponentWithEditor = withEditor<Props>(Component);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });

  const useToggleEditorFnNoOpen = vi.fn().mockReturnValue({
    closeEditor: vi.fn(),
    openEditor: vi.fn(),
    isOpenEditor: false,
  });
  it("Shouldn't render Component with isOpenEditor false", () => {
    const ComponentWithEditor = withEditor<Props>(Component, {}, useToggleEditorFnNoOpen);
    const { asFragment } = render(<ComponentWithEditor className="af-component" onChange={onChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('useToggleEditor', () => {
  it('Should isOpenEditor to be false when closeEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.closeEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(false);
  });

  it('Should isOpenEditor to be true when openEditor called', () => {
    const { result } = renderHook(() => useToggleEditor());
    act(() => result.current.openEditor());
    const resIsOpenEditor = result.current.isOpenEditor;
    expect(resIsOpenEditor).toBe(true);
  });
});

describe('LabelEditor', () => {
  it('Should render label when value is not a separator', () => {
    const { asFragment } = render(<LabelEditor value="value" name="name" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render nothing when value is a separator', () => {
    const { asFragment } = render(<LabelEditor value={{ value: 'sep', type: 'separator' }} name="name" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
