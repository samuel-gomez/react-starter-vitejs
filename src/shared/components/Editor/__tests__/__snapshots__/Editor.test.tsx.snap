// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LabelEditor > Should render label when value is not a separator 1`] = `
<DocumentFragment>
  <label
    class="af-form__group-label"
    for="name"
  >
    name
  </label>
</DocumentFragment>
`;

exports[`LabelEditor > Should render nothing when value is a separator 1`] = `<DocumentFragment />`;

exports[`withEditor > Should render Component with FormEditor when apply withEditor HOC 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;

exports[`withEditor > Should render Component with FormEditor when apply withEditor HOC and isOpenEditor true 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
    <div
      aria-label="Editor props"
      class="af-draggable-container react-draggable"
      style="transform: translate(0px,0px);"
    >
      <div
        class="af-draggable"
      >
        <h3
          class="af-draggable__title"
        >
          Props Editor
          <div
            class="af-draggable__tools"
          >
            <i
              class="glyphicon glyphicon-move"
            />
            <button
              aria-label="close Editor"
              class="af-btn--circle af-btn--circle--close-editor"
              type="button"
            >
              <i
                class="glyphicon glyphicon-close"
              />
            </button>
          </div>
        </h3>
        <div
          class="af-draggable__content"
        >
          <form
            class="af-form-editor"
          >
            <div
              class="af-form-editor__field"
            >
              <label
                class="af-form__group-label"
                for="className"
              >
                className
              </label>
              <input
                class="af-form__input-text"
                id="className"
                name="className"
                type="text"
                value="af-component"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</DocumentFragment>
`;

exports[`withEditor > Shouldn't render Component with isOpenEditor false 1`] = `
<DocumentFragment>
  <section
    class="af-editor"
  >
    <input
      class="af-component"
      id="idtext"
      name="nametext"
      type="text"
      value="hello"
    />
  </section>
</DocumentFragment>
`;
